### PR TITLE
Extend CSI Snapshot support to Longhorn snapshot

### PR DESCRIPTION
**Pull request by PhanLe1010**
longhorn/longhorn#2534

### CODE REVIEW

File: csi/controller_server.go [hunk 1]
```diff
@@ -623,53 +708,59 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
...
-// decodeSnapshotID splits up the snapshotID back into it's components
-// backupType will be used once we implement the VolumeBackup crd
-func decodeSnapshotID(snapshotID string) (backupType, volumeName, backupName string) {
+func decodeSnapshotID(snapshotID string) (csiSnapshotType, sourceVolumeName, id string) {
 	split := strings.Split(snapshotID, "://")
 	if len(split) < 2 {
 		return "", "", snapshotID
 	}
-	backupType = split[0]
+	if split[0] == "bs" {
```
Comment (shuo-wu) [inline csi/controller_server.go:753]: NIT: Please create constants for `snap` and `bs`. e.g., `SnapshotIDPrefixLonghornSnapshot`, `SnapshotIDPrefixLonghornBackup`

File: csi/controller_server.go [hunk 2]
```diff
@@ -678,17 +769,32 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
...
-
-	if backupVolume != nil && backupVolume.Name != "" {
-		backupVolume, err = cs.apiClient.BackupVolume.ActionBackupDelete(backupVolume, &longhornclient.BackupInput{Name: backupName})
+	csiSnapshotType, sourceVolumeName, id := decodeSnapshotID(snapshotID)
+	if csiSnapshotType == csiSnapshotTypeLonghornSnapshot {
+		volume, err := cs.apiClient.Volume.ById(sourceVolumeName)
 		if err != nil {
 			return nil, status.Error(codes.Internal, err.Error())
 		}
+		if volume != nil {
```
Comment (shuo-wu) [inline csi/controller_server.go:778]: THe volume should be `attached`.

**Review by shuo-wu — COMMENTED** [review]
In general LGTM

File: csi/controller_server.go [hunk 3]
```diff
@@ -537,6 +548,79 @@ func (cs *ControllerServer) GetCapacity(context.Context, *csi.GetCapacityRequest
...
+	}
+
+	vol, err := cs.apiClient.Volume.ById(csiVolumeName)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if vol == nil {
+		return nil, status.Errorf(codes.NotFound, "volume %s not found", csiVolumeName)
+	}
+	if vol.State != string(longhorn.VolumeStateAttached) {
```
[the remark this answers is not in the document]
Comment (innobead) [reply csi/controller_server.go:592]: We should as per the CSI sanity testing.

File: k8s/pkg/apis/longhorn/v1beta1/volume.go [hunk 4]
```diff
@@ -31,14 +31,14 @@ const (
 	VolumeFrontendEmpty    = VolumeFrontend("")
 )
 
-// +kubebuilder:validation:Enum="";backup;snapshot;volume
 type VolumeDataSource string
 
+type VolumeDataSourceType string
+
 const (
-	VolumeDataSourceTypeNone     = VolumeDataSource("")
```
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply k8s/pkg/apis/longhorn/v1beta1/volume.go]: Thank you

File: csi/controller_server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply csi/controller_server.go:592]: Thank you. Added

File: csi/controller_server.go [hunk 5]
```diff
@@ -537,6 +548,79 @@ func (cs *ControllerServer) GetCapacity(context.Context, *csi.GetCapacityRequest
...
+	}
+
+	var snapshot *longhornclient.Snapshot
+	snapshotListOutput, err := cs.apiClient.Volume.ActionSnapshotList(vol)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	for _, snap := range snapshotListOutput.Data {
+		if snap.Name == csiSnapshotName {
+			snapshot = &snap
```
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply csi/controller_server.go:603]: Thank you. Fixed

File: csi/controller_server.go (hunk 1 above)
Comment (PhanLe1010) [reply csi/controller_server.go:753]: Thank you. Fixed

File: csi/controller_server.go (hunk 2 above)
Comment (PhanLe1010) [reply csi/controller_server.go:778]: If the volume is not attached. API will return error

**Review by innobead — APPROVED** [review]
LGTM

**Review by joshimoo — APPROVED** [review]
Can be merged, left some comments.
I think using the same type schema for internal/external would make the code easier :)
Since the snapshot class param doesn't need to be user friendly, it's part of an internal representation, and users need to read the documentation anyway to know what types are available :)

File: csi/controller_server.go [hunk 6]
```diff
@@ -623,53 +712,59 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
...
+		// when trying to find a backup for deletion or restoration
+		return fmt.Sprintf("%s://%s/%s", snapshotIDPrefixLonghornBackup, sourceVolumeName, id)
+	}
+	return "INVALID-ID"
 }
 
-// decodeSnapshotID splits up the snapshotID back into it's components
-// backupType will be used once we implement the VolumeBackup crd
-func decodeSnapshotID(snapshotID string) (backupType, volumeName, backupName string) {
+func decodeSnapshotID(snapshotID string) (csiSnapshotType, sourceVolumeName, id string) {
```
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply csi/controller_server.go:748]: Thank you. fixed

File: csi/controller_server.go [hunk 7]
```diff
@@ -31,6 +31,11 @@ const (
 	timeoutBackupInitiation = 60 * time.Second
 	tickBackupInitiation    = 5 * time.Second
 	backupStateCompleted    = "Completed"
+
+	csiSnapshotTypeLonghornSnapshot  = "longhorn-snapshot"
+	csiSnapshotTypeLonghornBackup    = "longhorn-backup"
+	snapshotIDPrefixLonghornSnapshot = "snap"
+	snapshotIDPrefixLonghornBackup   = "bs"
```
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply csi/controller_server.go]: Thank you for the advice. Fixed

File: csi/controller_server.go [hunk 8]
```diff
@@ -537,6 +550,81 @@ func (cs *ControllerServer) GetCapacity(context.Context, *csi.GetCapacityRequest
...
+		rsp, err = cs.createCSISnapshotTypeLonghornSnapshot(req)
+	} else if csiSnapshotType == "" || csiSnapshotType == csiSnapshotTypeLonghornBackup {
+		// For backward compatibility, empty type is considered as csiSnapshotTypeLonghornBackup
+		rsp, err = cs.createCSISnapshotTypeLonghornBackup(req)
+	} else {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid snapshot type: %v. Must be %v or %v or \"\"", csiSnapshotType, csiSnapshotTypeLonghornSnapshot, csiSnapshotTypeLonghornBackup)
+	}
+
+	return rsp, err
+}
```
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply csi/controller_server.go]: Thank you. Yeah, agree. We should move the logic into backend API

**Review by joshimoo — DISMISSED** [review]

**Review by joshimoo — APPROVED** [review]
LGTM is good enough for now, in a future refactor the business/attachment logic should be handled by a backend call. Just leaving a REF for that in the comment https://github.com/longhorn/longhorn-manager/pull/1037

**Pull request merged**